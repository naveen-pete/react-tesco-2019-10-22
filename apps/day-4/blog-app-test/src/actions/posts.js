import * as PostActionType from './index';
import * as BookAppApi from '../api';

export function getPosts() {
  const request = BookAppApi.getPosts();

  return {
    type: PostActionType.GET_POSTS,
    payload: request
  };
}

export function getPost(id, callback = null) {
  const request = BookAppApi.getPost(id);

  if (callback) {
    request.then(callback);
  }

  return {
    type: PostActionType.GET_POST,
    payload: request
  };
}

export function createPost(post, callback = null) {
  const request = BookAppApi.addPost(post);

  if (callback) {
    request.then(callback);
  }

  return {
    type: PostActionType.CREATE_POST,
    payload: request
  };
}

export function updatePost(post, callback = null) {
  const request = BookAppApi.updatePost(post);

  if (callback) {
    request.then(callback);
  }

  return {
    type: PostActionType.UPDATE_POST,
    payload: request
  };
}

export function deletePost(id, callback = null) {
  const request = BookAppApi.deletePost(id);

  if (callback) {
    request.then(callback);
  }

  return {
    type: PostActionType.DELETE_POST,
    payload: id
  };
}

export function getCategories() {
  const request = BookAppApi.getCategories();

  return {
    type: PostActionType.GET_CATEGORIES,
    payload: request
  };
}

export function setCurrentCategory(category) {
  return {
    type: PostActionType.SET_CURRENT_CATEGORY,
    payload: category
  };
}
