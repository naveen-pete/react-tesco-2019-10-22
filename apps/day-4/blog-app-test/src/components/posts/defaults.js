export const CATEGORY_ALL = { code: 'all', name: 'All' };
