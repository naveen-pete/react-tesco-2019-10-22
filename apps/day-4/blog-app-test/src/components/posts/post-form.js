import React, { Component } from 'react';
import FontAwesome from 'react-fontawesome';
import { Link } from 'react-router-dom';

import { connect } from 'react-redux';

import {
  getCategories,
  getPost,
  createPost,
  updatePost
} from '../../actions/posts';

const defaultPost = {
  title: '',
  body: '',
  author: '',
  category: '',
  commentCount: 0,
  voteScore: 0,
  timestamp: 0
};

export class PostForm extends Component {
  constructor() {
    super();

    this.state = {
      post: defaultPost
    };

    this.onFormSubmit = this.onFormSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentDidMount() {
    this.props.getCategories();
    const id = this.props.match.params.id;
    if (id) {
      this.props.getPost(id, post => {
        this.setState({ post });
      });
    }
  }

  onFormSubmit(event) {
    event.preventDefault();

    // #1: older style - using Object.assign() method
    // const post = Object.assign({}, this.state.post);

    // #2: newer style - using object destructuring syntax
    const post = { ...this.state.post };
    const callback = () => this.props.history.push('/');

    if (post.id) {
      this.props.updatePost(post, callback);
    } else {
      post.timestamp = Date.now();
      this.props.createPost(post, callback);
    }
  }

  handleChange(event) {
    // Retrieve name, value and control type flag of the form control
    const { name, value } = event.target;

    this.setState(prevState => {
      // Assign the updated value
      // #1: older style - using Object.assign() method
      // const newPost = Object.assign({}, prevState.post, { [name]: value });

      // #2: newer style - using object destructuring syntax
      const newPost = {
        ...prevState.post,
        [name]: value
      };

      return { post: newPost };
    });
  }

  render() {
    const { post } = this.state;
    const { categories } = this.props;

    return (
      <div className="row">
        <div className="col-sm-12">
          <h4>
            <FontAwesome name="envelope" /> Post Form
          </h4>
          <div className="well well-lg">
            <form onSubmit={this.onFormSubmit}>
              <div className="form-group">
                <label htmlFor="title">Title</label>
                <input
                  required
                  type="text"
                  className="form-control"
                  id="title"
                  name="title"
                  value={post.title}
                  onChange={this.handleChange}
                />
              </div>

              <div className="form-group">
                <label htmlFor="body">Body</label>
                <textarea
                  required
                  className="form-control"
                  id="body"
                  name="body"
                  cols="30"
                  rows="4"
                  value={post.body}
                  onChange={this.handleChange}
                />
              </div>

              <div className="form-group">
                <label htmlFor="author">Author</label>
                <input
                  required
                  type="text"
                  className="form-control"
                  id="author"
                  name="author"
                  value={post.author}
                  onChange={this.handleChange}
                />
              </div>

              <div className="form-group">
                <label htmlFor="category">Category</label>
                <select
                  required
                  className="form-control"
                  id="category"
                  name="category"
                  value={post.category}
                  onChange={this.handleChange}
                >
                  <option value="" />
                  {categories.map(({ code, name }) => (
                    <option key={code} value={code}>
                      {name}
                    </option>
                  ))}
                </select>
              </div>

              <button
                type="submit"
                className="btn btn-primary btn-sm smallMargin"
              >
                <span className="glyphicon glyphicon-ok" /> Save
              </button>
              <Link to="/" className="btn btn-warning btn-sm smallMargin">
                <span className="glyphicon glyphicon-remove" /> Cancel
              </Link>
            </form>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ categories }, props) => ({
  categories
});

const mapDispatchToProps = dispatch => ({
  getCategories: () => dispatch(getCategories()),
  getPost: (id, callback) => dispatch(getPost(id, callback)),
  createPost: (post, callback) => dispatch(createPost(post, callback)),
  updatePost: (post, callback) => dispatch(updatePost(post, callback))
});

export default connect(mapStateToProps, mapDispatchToProps)(PostForm);
