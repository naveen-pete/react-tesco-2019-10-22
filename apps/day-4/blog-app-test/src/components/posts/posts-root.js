import React from 'react';

import Categories from './categories';
import PostList from './post-list';

// -------------------------------------------------------
// The root component for Posts module. It uses:
// 1. Categories component to show the list of categories
// 2. PostList component to manage list of posts
// -------------------------------------------------------
const PostsRoot = () => {
  return (
    <div className="row">
      <Categories />
      <PostList />
    </div>
  );
};

export default PostsRoot;
