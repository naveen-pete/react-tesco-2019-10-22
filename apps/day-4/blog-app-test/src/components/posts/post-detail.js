import React, { Component } from 'react';
import FontAwesome from 'react-fontawesome';
import { Link } from 'react-router-dom';

import { connect } from 'react-redux';

import AppAlert from '../common/app-alert';

import { getPost, deletePost } from '../../actions/posts';

class PostDetail extends Component {
  componentDidMount() {
    const id = this.props.match.params.id;
    this.props.getPost(id);
  }

  deletePost(id) {
    this.props.deletePost(id, () => this.props.history.push('/'));
  }

  render() {
    const { post } = this.props;

    if (!post) {
      return <AppAlert type="info" message="Loading..." />;
    }

    return (
      <div className="row">
        <div className="col-sm-12">
          <div className="row">
            <div className="col-sm-12">
              <h4>
                <FontAwesome name="envelope" /> Post Detail
              </h4>
            </div>
          </div>
          <div className="row">
            <div className="col-sm-12">
              <Link to="/" className="btn btn-sm btn-primary smallMargin">
                <span className="glyphicon glyphicon-chevron-left" /> Posts
              </Link>
            </div>
          </div>
          <br />
          <div className="row">
            <div className="col-sm-12">
              <div className="panel panel-info">
                <div className="panel-heading">
                  <h3 className="panel-title">
                    <FontAwesome name="envelope" /> {post.title}
                  </h3>
                </div>
                <div className="panel-body">
                  <p>{post.body}</p>
                </div>
                <ul className="list-group">
                  <li className="list-group-item">
                    <FontAwesome name="user" /> Author: {post.author}
                  </li>
                  <li className="list-group-item">
                    <FontAwesome name="object-group" /> Category:{' '}
                    {post.category}
                  </li>
                  <li className="list-group-item">
                    <FontAwesome name="comments-o" /> {post.commentCount}{' '}
                    comment(s)
                  </li>
                  <li className="list-group-item">{post.voteScore} like(s)</li>
                </ul>
                <div className="panel-footer">
                  <Link
                    to={`/posts/${post.id}/edit`}
                    className="btn btn-primary btn-sm smallMargin"
                  >
                    <span className="glyphicon glyphicon-edit" /> Edit
                  </Link>
                  <button
                    onClick={() => this.deletePost(post.id)}
                    className="btn btn-warning btn-sm smallMargin"
                  >
                    <span className="glyphicon glyphicon-trash" /> Delete
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ posts }, props) => ({
  post: posts.find(post => post.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  getPost: id => dispatch(getPost(id)),
  deletePost: (id, callback) => dispatch(deletePost(id, callback))
});

export default connect(mapStateToProps, mapDispatchToProps)(PostDetail);
