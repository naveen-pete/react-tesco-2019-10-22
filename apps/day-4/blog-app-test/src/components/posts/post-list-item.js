import React from 'react';
import FontAwesome from 'react-fontawesome';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { getFormattedDate } from '../../helpers';

import { deletePost } from '../../actions/posts';

const PostListItem = props => {
  const { post } = props;

  return (
    <div className="panel panel-info">
      <div className="panel-heading">
        <FontAwesome name="envelope" /> {post.title}
      </div>
      <div className="panel-body">
        <div className="row">
          <div className="col-sm-8">
            <FontAwesome name="user" /> {post.author}
          </div>
          <div className="col-sm-4">
            <FontAwesome name="object-group" /> {post.category}
          </div>
        </div>
        <div className="row">
          <div className="col-sm-4">
            <FontAwesome name="comments-o" /> {post.commentCount} comment(s)
          </div>
          <div className="col-sm-4">{post.voteScore} like(s)</div>
          <div className="col-sm-4">
            <FontAwesome name="clock-o" /> {getFormattedDate(post.timestamp)}
          </div>
        </div>
      </div>
      <div className="panel-footer">
        <Link
          to={`/posts/${post.id}`}
          className="btn btn-xs btn-info smallMargin"
        >
          <FontAwesome name="envelope-open-o" /> View
        </Link>
        <Link
          to={`/posts/${post.id}/edit`}
          className="btn btn-xs btn-primary smallMargin"
        >
          <span className="glyphicon glyphicon-edit" /> Edit
        </Link>
        <button
          onClick={() => props.deletePost(post.id)}
          className="btn btn-xs btn-warning smallMargin"
        >
          <span className="glyphicon glyphicon-trash" /> Delete
        </button>
      </div>
    </div>
  );
};

const mapDispatchToProps = dispatch => ({
  deletePost: id => dispatch(deletePost(id))
});

export { PostListItem };
export default connect(null, mapDispatchToProps)(PostListItem);
