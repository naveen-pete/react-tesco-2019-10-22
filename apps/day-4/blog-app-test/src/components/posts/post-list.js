import React, { Component } from 'react';
import FontAwesome from 'react-fontawesome';
import { Link } from 'react-router-dom';

import { connect } from 'react-redux';

import PostListItem from './post-list-item';
import AppAlert from '../common/app-alert';

import { getPosts } from '../../actions/posts';

export class PostList extends Component {
  componentDidMount() {
    this.props.getPosts();
  }

  render() {
    const { posts, currentCategory } = this.props;

    const filteredPosts =
      currentCategory.code === 'all'
        ? posts
        : posts.filter(post => post.category === currentCategory.code);

    return (
      <div className="col-sm-9">
        <div className="pull-right">
          <Link to="/posts/new" className="btn btn-sm btn-primary">
            <span className="glyphicon glyphicon-plus" /> New Post
          </Link>
        </div>
        <h4>
          <FontAwesome name="envelope" /> Posts (Category:{' '}
            <span id="category">{currentCategory.name}</span>)
        </h4>
        {filteredPosts.length > 0 ? (
          filteredPosts.map(post => <PostListItem key={post.id} post={post} />)
        ) : (
          <AppAlert
            type="info"
            message="No posts available for selected category."
          />
        )}
      </div>
    );
  }
}

const mapStateToProps = ({ posts, currentCategory }) => ({
  posts,
  currentCategory
});

const mapDispatchToProps = dispatch => ({
  getPosts: () => dispatch(getPosts())
});

export default connect(mapStateToProps, mapDispatchToProps)(PostList);
