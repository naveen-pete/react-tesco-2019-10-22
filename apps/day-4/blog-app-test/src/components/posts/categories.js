import React, { Component } from 'react';
import FontAwesome from 'react-fontawesome';

import { connect } from 'react-redux';

import { CATEGORY_ALL } from './defaults';
import { getCategories, setCurrentCategory } from '../../actions/posts';

export class Categories extends Component {
  componentDidMount() {
    this.props.getCategories();
  }

  onCategoryClick(selectedCategory) {
    this.props.setCurrentCategory(selectedCategory);
  }

  addActiveIfSelected(category) {
    const { code } = this.props.currentCategory;
    return code === category.code ? 'active' : '';
  }

  render() {
    let { categories } = this.props;
    categories = [CATEGORY_ALL, ...categories];

    return (
      <div className="col-sm-3">
        <h4>
          <FontAwesome name="object-group" /> Category
        </h4>
        <ul className="list-group">
          {categories.map(category => (
            <li
              onClick={() => this.onCategoryClick(category)}
              key={category.code}
              className={`list-group-item cursorPointer ${this.addActiveIfSelected(category)}`}
            >
              {category.name}
            </li>
          ))}
        </ul>
      </div>
    );
  }
}

const mapStateToProps = ({ categories, currentCategory }) => ({
  categories,
  currentCategory
});

const mapDispatchToProps = dispatch => ({
  getCategories: () => dispatch(getCategories()),
  setCurrentCategory: category => dispatch(setCurrentCategory(category))
});

export default connect(mapStateToProps, mapDispatchToProps)(Categories);
