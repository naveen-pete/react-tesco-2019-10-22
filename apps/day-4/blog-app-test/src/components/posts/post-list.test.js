import React from 'react';
import { shallow } from 'enzyme';
import { PostList } from './post-list';
import posts from '../../test-fixtures/posts';
import { CATEGORY_ALL } from './defaults';

test('should render posts', () => {
  const wrapper = shallow(<PostList posts={posts} currentCategory={CATEGORY_ALL} getPosts={() => {}} />);
  expect(wrapper.find('Connect(PostListItem)')).toHaveLength(posts.length);
  expect(wrapper).toMatchSnapshot();
});

test('should render a message when posts array is empty', () => {
  const wrapper = shallow(<PostList posts={[]} currentCategory={CATEGORY_ALL} getPosts={() => {}} />);
  expect(wrapper.find('AppAlert')).toHaveLength(1);
  expect(wrapper).toMatchSnapshot();
});

test('should render posts for a particular category', () => {
  const category = {
    code: 'redux',
    name: 'Redux'
  }
  const wrapper = shallow(<PostList posts={posts} currentCategory={category} getPosts={() => {}} />);
  expect(wrapper.find('Connect(PostListItem)')).toHaveLength(2);
  expect(wrapper).toMatchSnapshot();
});

test('should render the name of category passed as a prop', () => {
  const category = {
    code: 'redux',
    name: 'Redux'
  }
  const wrapper = shallow(<PostList posts={posts} currentCategory={category} getPosts={() => {}} />);
  expect(wrapper.find('#category').text()).toBe(category.name);
});

test('should retrieve posts when the component is mounted', () => {
  const getPostsSpy = jest.fn();
  const wrapper = shallow(
    <PostList 
      posts={posts} 
      currentCategory={CATEGORY_ALL} 
      getPosts={getPostsSpy} 
    />
  );

  expect(getPostsSpy).toBeCalled();
});
