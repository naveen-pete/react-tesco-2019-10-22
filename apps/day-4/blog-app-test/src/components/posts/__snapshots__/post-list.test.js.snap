// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a message when posts array is empty 1`] = `
<div
  className="col-sm-9"
>
  <div
    className="pull-right"
  >
    <Link
      className="btn btn-sm btn-primary"
      replace={false}
      to="/posts/new"
    >
      <span
        className="glyphicon glyphicon-plus"
      />
       New Post
    </Link>
  </div>
  <h4>
    <FontAwesome
      name="envelope"
    />
     Posts (Category:
     
    <span
      id="category"
    >
      All
    </span>
    )
  </h4>
  <AppAlert
    message="No posts available for selected category."
    type="info"
  />
</div>
`;

exports[`should render posts 1`] = `
<div
  className="col-sm-9"
>
  <div
    className="pull-right"
  >
    <Link
      className="btn btn-sm btn-primary"
      replace={false}
      to="/posts/new"
    >
      <span
        className="glyphicon glyphicon-plus"
      />
       New Post
    </Link>
  </div>
  <h4>
    <FontAwesome
      name="envelope"
    />
     Posts (Category:
     
    <span
      id="category"
    >
      All
    </span>
    )
  </h4>
  <Connect(PostListItem)
    key="1"
    post={
      Object {
        "author": "facebook",
        "body": "Redux can be described in three fundamental principles: (a) Single source of truth (b) State is read-only (c) Changes are made with pure functions",
        "category": "redux",
        "commentCount": 10,
        "id": 1,
        "timestamp": 1468479767290,
        "title": "Redux Principles",
        "voteScore": 4,
      }
    }
  />
  <Connect(PostListItem)
    key="2"
    post={
      Object {
        "author": "google",
        "body": "One framework. Mobile & desktop.",
        "category": "angular",
        "commentCount": 15,
        "id": 2,
        "timestamp": 1468479767290,
        "title": "Angular",
        "voteScore": 6,
      }
    }
  />
  <Connect(PostListItem)
    key="3"
    post={
      Object {
        "author": "thingone",
        "body": "Just kidding. It takes more than 10 minutes to learn technology.",
        "category": "redux",
        "commentCount": 20,
        "id": 3,
        "timestamp": 1468479767190,
        "title": "Learn Redux in 10 minutes!",
        "voteScore": 8,
      }
    }
  />
</div>
`;

exports[`should render posts for a particular category 1`] = `
<div
  className="col-sm-9"
>
  <div
    className="pull-right"
  >
    <Link
      className="btn btn-sm btn-primary"
      replace={false}
      to="/posts/new"
    >
      <span
        className="glyphicon glyphicon-plus"
      />
       New Post
    </Link>
  </div>
  <h4>
    <FontAwesome
      name="envelope"
    />
     Posts (Category:
     
    <span
      id="category"
    >
      Redux
    </span>
    )
  </h4>
  <Connect(PostListItem)
    key="1"
    post={
      Object {
        "author": "facebook",
        "body": "Redux can be described in three fundamental principles: (a) Single source of truth (b) State is read-only (c) Changes are made with pure functions",
        "category": "redux",
        "commentCount": 10,
        "id": 1,
        "timestamp": 1468479767290,
        "title": "Redux Principles",
        "voteScore": 4,
      }
    }
  />
  <Connect(PostListItem)
    key="3"
    post={
      Object {
        "author": "thingone",
        "body": "Just kidding. It takes more than 10 minutes to learn technology.",
        "category": "redux",
        "commentCount": 20,
        "id": 3,
        "timestamp": 1468479767190,
        "title": "Learn Redux in 10 minutes!",
        "voteScore": 8,
      }
    }
  />
</div>
`;
